README.txt
= entropy-gem

This small gem provides methods to calculate the 
entropy (and cardinality and diversity) of a probability
space and of a probability metric space.

A probability space is defined by a probability
vector [p1, p2, ..., pn]. A probability metric space 
is defined by a probability vector and a metric function.
The metric function defines the "distance" between the point
i and the point j along with a distance matrix D where
D[i, j] = distance between i and j.

Most of the code is derived from:
http://golem.ph.utexas.edu/category/2008/10/entropy_diversity_and_cardinal.html
and 
http://golem.ph.utexas.edu/category/2008/11/entropy_diversity_and_cardinal_1.html

In order to define a probability space, you can use the function add_stream to
add a stream of chars (or bit). The method compute the probability
vector using the frequency of each element of the stream. Otherwise
you could supply directly a probability vector as an array of reals.

For a probability metric space you should also define the distance metric,
either passing your function as a block to define_distance, or passing
a matrix to the same define_distance.

The example files  show a basic usage of the classes.

NOTE: requires ruby >= 1.9.x

Feel free to write me: rubygems@hiraedd.com or twitter @hiraedd
if you have any question or suggestion.

== Copyright

Copyright (c) 2011 hiraedd@hiraedd.com. See LICENSE.txt for further details.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.



CHANGELOG.txt
== Version 0.1.0

* Add a test to verify a metric space degenerating in 
  discrete case (distance = Infinity) and modified the examples

== Version 0.0.3

*  Specify more clearly that requires ruby 1.9.x

== Version 0.0.2

* Specifiy the requirements of ruby 1.9.x

== Version 0.0.1

* Initial commit



test/test_entropy_def.rb
require 'test/unit'
require 'entropy'

class TestEntropy < Test::Unit::TestCase

  def test_version_string
    assert_equal Entropy.version_string, "Entropy version #{Entropy::VERSION}"
  end
end



test/test_entropy_metr.rb
require 'test/unit'
require 'entropy'

class TestEntropy < Test::Unit::TestCase

  def test_string_metr
    obj = Entropy::ProbabilityMetricSpace.new
    obj.add_stream("000000000")
    obj.define_distance {|i, j| Math.sqrt(i*i + j*j) }
    assert_equal obj.prob_space, [1.0]
  end
  
  def test_string_metr_2
    obj = Entropy::ProbabilityMetricSpace.new
    obj.add_stream("1000")
    obj.define_distance {|i, j| Math.sqrt(i*i + j*j) }
    assert_equal obj.prob_space.sort, [0.25, 0.75]
  end

  def test_distance_metr_3
    obj = Entropy::ProbabilityMetricSpace.new
    obj.add_stream("0123")
    obj.define_distance {|i, j| Math.sqrt(i*i + j*j) }
    m = Matrix.build(4, 4) {|i, j| Math.sqrt(i*i + j*j) }
    assert_equal obj.distance_matrix, m
  end

  def test_entropy_metr_4
    obj = Entropy::ProbabilityMetricSpace.new
    obj.add_stream("0123")
    obj.define_distance {|i, j| 1 }
    assert_equal obj.diversity(1), 1
  end

  def test_entropy_metr_5
    obj = Entropy::ProbabilityMetricSpace.new
    obj.add_stream("01")
    m = Matrix[[0, :infinite], [:infinite, 0]]
    obj.define_distance(m)
    assert_equal obj.cardinality(2), 2
    assert_equal obj.cardinality(42), 2
  end
end



test/test_entropy_prob.rb
require 'test/unit'
require 'entropy'

class TestEntropy < Test::Unit::TestCase

  def test_string_prob
    obj = Entropy::ProbabilitySpace.new
    obj.add_stream("000000000")
    assert_equal obj.prob_space, [1.0]
  end
  
  def test_string_prob_2
    obj = Entropy::ProbabilitySpace.new
    obj.add_stream("1000")
    assert_equal obj.prob_space.sort, [0.25, 0.75]
  end

  def test_cardinality 
    obj = Entropy::ProbabilitySpace.new
    stream = "11111111"
    obj.add_stream(stream)
    assert_equal obj.cardinality(2).round, 1
  end
  
  def test_cardinality_2
    obj = Entropy::ProbabilitySpace.new
    stream = "1234"
    obj.add_stream(stream)
    assert_equal obj.cardinality(2).round, stream.size
  end

  def test_diversity
    obj = Entropy::ProbabilitySpace.new
    stream = "11111111"
    obj.add_stream(stream)
    assert_equal obj.diversity(2).round, 0
    assert_equal obj.renyi_entropy(:infinite).round, 0
  end
  
  def test_diversity_2
    obj = Entropy::ProbabilitySpace.new
    stream = "1234"
    obj.add_stream(stream)
    assert_equal obj.diversity(2), obj.surprise(2, 1.0/stream.size)
    assert_equal obj.renyi_entropy(:infinite), -Math.log(0.25)
  end
  
  def test_entropy_renyi
    obj = Entropy::ProbabilitySpace.new
    stream = "11111111"
    obj.add_stream(stream)
    assert_equal obj.renyi_entropy(2), 0
  end

  def test_entropy_renyi_2
    obj = Entropy::ProbabilitySpace.new
    stream = "1234"
    obj.add_stream(stream)
    p = 0.25*0.25
    assert_equal obj.renyi_entropy(2), -Math.log(4*p)
    assert_equal obj.renyi_entropy(:infinite), -Math.log(0.25)
  end
end



example/prob_space.rb
require 'entropy'
require 'pp'

# note: require ruby 1.9.x

obj = Entropy::ProbabilitySpace.new
obj.add_stream("000000000")
pp obj.prob_space   # [1.0]
puts obj.cardinality(2) # 1 equal to shannon entropy
puts obj.diversity(2).round # 0
puts obj.renyi_entropy(2).round # 0
puts obj.renyi_entropy(:infinite).round # 0
  
obj = Entropy::ProbabilitySpace.new
obj.add_stream("1234")
pp obj.prob_space # [0.25, 0.25, 0.25, 0.25]
puts obj.cardinality(2) # 4 equal to shannon entropy
puts obj.diversity(2) # 3/4
puts obj.renyi_entropy(2) # 2 
puts obj.renyi_entropy(:infinite) 

# redefine prob_space
puts "redefine prob_space"
obj.prob_space = [0.25, 0.75]
pp obj.prob_space # [0.25, 0.25, 0.25, 0.25]
puts obj.cardinality(2) # 1.6
puts obj.diversity(2) # 0.375
puts obj.renyi_entropy(2) 
puts obj.renyi_entropy(0) 
puts obj.renyi_entropy(:infinite) 





lib/entropy/stream.rb
module Entropy

  # Class used to define a stream inside the space
  class Stream

    attr_reader :streams

    def initialize
      @streams = Hash.new
    end

    # Adds a stream (a File or a String) to the space
    def add_stream(stream)
      iter = nil 
      if stream.respond_to?("binmode?")
        if stream.binmode?
          puts "Binary File Stream Initialization..."
          iter = :each_byte
        else
          puts "Char File Stream Initialization..."
          iter = :each_char
        end
      elsif  stream.respond_to?("each_char")
        puts "Char Stream Initialization..."
        iter = :each_char
      elsif  stream.respond_to?("each_byte")
        puts "Binary Stream Initialization..."
        iter = :each_byte
      else
        raise "Wrong type of stream"
      end
      @streams[stream] = iter
    end

  end

end


lib/entropy/metric_space.rb
require "matrix"

module Entropy

  # Entropy::ProbabilityMetricSpace defines a finite probability metric space
  class ProbabilityMetricSpace < ProbabilitySpace

    attr_reader :distance_matrix, :distance_function

    def initialize
      @distance_matrix = nil
      @density = nil
      super
    end

    # Adds a set of elements (char, bit) to the space
    def add_stream(stream)
      super(stream)
    end

    # Defines the distance among the elements of the set.
    # The method accepts a block, so you can use your
    # function to calculate the distance. The function should
    # accept two parameters, viz. the ordinals i,j of the elements
    # and should return a decimal number, the distance between 
    # element i and j.
    # In alternative you could pass a matrix of numbers defining
    # the distances among i's and j's.
    def define_distance(*args, &blk)
      if !blk.nil?
        @distance_function = blk
        update_distance_matrix
      else
        mat = args[0]
        raise "distance matrix is not a matrix object" if !mat.respond_to?("square?") 
        raise "distance matrix is not square" if !mat.square?
        raise "distance matrix has wrong dimensions" if mat.to_a.size != @prob_space.size
        @distance_matrix = mat
      end
      z = @distance_matrix.map { |d| d == :infinite ? 0 : Math.exp(-d) }
      @density = z * Matrix.column_vector(@prob_space)
      @density = @density.column(0).to_a
    end

    # Computes the Shannon Entropy of the space
    def shannon_entropy
      diversity(1)
    end

    # Computes the Rao Entropy of the space
    def rao_entropy
      diversity(2)
    end


    # Computes the Renyi Entropy (α-Entropy) of the space
    def renyi_entropy(alpha)
      raise "Warning: distance metric not defined..." if @distance_matrix.nil?
      h = 0
        if alpha == 1
          @prob_space.each_index do |i| 
            h += -@prob_space[i] * Math.log(@density[i])
          end
        elsif alpha == :infinite
          h = -Math.log(@density.max)
        else
          @prob_space.each_index do |i| 
            h += @prob_space[i] * (@density[i]**(alpha - 1))
          end
          h = 1.0 / (1 - alpha) * Math.log(h)
        end
      h
    end

    # Computes the α-Diversity of the space
    def diversity(alpha)
      raise "Warning: distance metric not defined..." if @distance_matrix.nil?
      div = []
      @prob_space.each_index do |i| 
        p = @prob_space[i]
        div[i] = p *  surprise(alpha, @density[i])
      end
      div.inject(:+)
    end
    
    # Computes the α-Cardinality of the space
    def cardinality(alpha)
      raise "Warning: distance metric not defined..." if @distance_matrix.nil?
      card = 0
      if alpha == 1
        card = 1
        @prob_space.each_index do |i| 
          card = card * (@density[i] ** (-@prob_space[i]))
        end
      elsif alpha == :infinite
        card = 1.0/@density.max
      else
        @prob_space.each_index do |i| 
          card += @prob_space[i] * (@density[i]**(alpha - 1))
        end
        card = card ** (1.0 / (1 - alpha)) if card != 0
      end
      card
    end

    private
    def update_distance_matrix
      if !@distance_function.nil?
        puts "Applying function #{@distance_function.to_s} to the probability space..."
        @distance_matrix = Matrix.build (@prob_space.size) { |i, j| @distance_function.call(i, j) }
      end
    end

  end

end


lib/entropy/probability_space.rb
module Entropy

  # Entropy::ProbabilitySpace defines a finite probability space
  class ProbabilitySpace

    attr_accessor :prob_space

    def initialize
      @space = Entropy::Stream.new
      @dict = Entropy::Dict.new()
      @prob_space = []
    end

    # Adds a set of elements (char, bit) to the space
    def add_stream(stream)
      puts "Adding a set in the probability space..."
      @space.add_stream(stream)
      update_probability_space(stream)
    end

    # Computes the Shannon Entropy of the space
    def shannon_entropy
      diversity(1)
    end

    # Computes the Renyi Entropy (α-Entropy) of the space
    def renyi_entropy(alpha)
      h = 0
        if alpha == 1
          h = @prob_space.map {|p| -p * Math.log(p)}.inject(:+)
        elsif alpha == :infinite
          h = -Math.log(@prob_space.max)
        else
          h =  @prob_space.map {|p| p**alpha}.inject(:+)
          h = 1.0 / (1 - alpha) * Math.log(h)
        end
      h
    end

    # Computes the α-Diversity of the space
    def diversity(alpha)
      @prob_space.map {|p| p * surprise(alpha, p) }.inject(:+)
    end
    
    # Surprise function
    def surprise(alpha, p)
      s = 0
      if alpha == 1
        s = -Math.log(p)
      elsif alpha == :infinite
        s = 0
      else
        e = alpha - 1
        s = 1.0 / e * (1 - p**e)
      end
      s
    end

    # Computes the α-Cardinality of the space
    def cardinality(alpha)
      card = 0
      if alpha == 1
        card = @prob_space.map {|p| p**(-p)}.inject(1, :*)
      elsif alpha == :infinite
        card = 1.0/@prob_space.max
      else
        card = @prob_space.map {|p| p**alpha}.inject(:+)
        card = card ** (1.0 / (1 - alpha))
      end
      card
    end

    private
    def update_probability_space(stream)
      @prob_space = []
      it = @space.streams[stream]
      stream.send(it).each do |b|
        c = b
        if it == :each_char
            c.chomp!
            next if c.empty?
        end
        @dict.upd(c)
      end
      @dict.scan_keys.each do |k|
        @prob_space << @dict.freq(k)
      end
    end
  end

end


example/metric_prob_space.rb
require 'entropy'
require 'pp'

# note: require ruby 1.9.x

obj = Entropy::ProbabilityMetricSpace.new
obj.add_stream("000000000")
obj.define_distance {|i, j| Math.sqrt(i*i + j*j) }
pp obj.prob_space # [1.0]

obj = Entropy::ProbabilityMetricSpace.new
obj.add_stream("0123")
obj.define_distance {|i, j| Math.sqrt(i*i + j*j) }
m = Matrix.build(4, 4) {|i, j| Math.sqrt(i*i + j*j) }
pp obj.distance_matrix  # = m
pp m

# redefine prob_space
#
puts "redefine prob_space and distance matrix"
obj.prob_space = [0.25, 0.75]
distance = Matrix[[2, 2], [4, 4]]
obj.define_distance(distance)
pp obj.prob_space # [0.25, 0.25, 0.25, 0.25]
pp distance
puts obj.cardinality(2) 
puts obj.diversity(2) 
puts obj.renyi_entropy(2) 
puts obj.renyi_entropy(0) 
puts obj.renyi_entropy(:infinite) 

# setting infinite distance means having a discrete set
puts "setting infinite distance means having a discrete set"
distance = Matrix[[0, :infinite], [:infinite, 0]]
obj.define_distance(distance)
pp obj.prob_space # [0.25, 0.25, 0.25, 0.25]
pp distance
puts obj.cardinality(2) 
puts obj.diversity(2) 
puts obj.renyi_entropy(2) 
puts obj.renyi_entropy(0) 
puts obj.renyi_entropy(:infinite) 


